packages/deployments/contracts/contracts/messaging/connectors/SpokeConnector.sol

```diff
@@ -283,35 +308,107 @@ abstract contract SpokeConnector is Connector, ConnectorManager, MerkleTreeManag
   }
 
   /**
    * @notice Must be able to call the `handle` function on the BridgeRouter contract. This is called
    * on the destination domain to handle incoming messages.
+   *
+   * Proving:
+   * Calculates the expected inbound root from an origin chain given a leaf (message hash),
+   * the index of the leaf, and the merkle proof of inclusion (path). Next, we check to ensure that this
+   * calculated inbound root is included in the current aggregateRoot, given its index in the aggregator
+   * tree and the proof of inclusion.
+   *
+   * Processing:
+   * After all messages have been proven, we dispatch each message to Connext (BridgeRouter) for
+   * execution.
+   *
+   * @dev Currently, ALL messages in a given batch must path to the same shared inboundRoot, meaning they
+   * must all share an origin. See open TODO below for a potential solution to enable multi-origin batches.
    * @dev Intended to be called by the relayer at specific intervals during runtime.
+   * @dev Will record a calculated root as having been proven if we've already proven that it was included
+   * in the aggregateRoot.
+   *
+   * @param _proofs Batch of Proofs containing messages for proving/processing.
+   * @param _aggregatorPath Merkle path of inclusion for the inbound root.
+   * @param _aggregatorIndex Index of the inbound root in the aggregator's merkle tree in the hub.
    */
   function proveAndProcess(
-    bytes memory _message,
-    bytes32[32] calldata _proof,
-    uint256 _index
+    Proof[] calldata _proofs,
+    bytes32[32] calldata _aggregatorPath,
+    uint256 _aggregatorIndex
   ) external onlyUnpaused {
-    // 1. must prove existence of the given outbound root from destination domain
-    // 2. must prove the existence of the given message in the destination
-    // domain outbound root
-    require(prove(keccak256(_message), _proof, _index), "!prove");
-    // FIXME: implement proofs above before processing the message
-    process(_message);
+    // Sanity check: proofs are included.
+    require(_proofs.length > 0, "!proofs");
+
+    // Optimization: calculate the inbound root for the first message in the batch and validate that
+    // it's included in the aggregator tree. We can use this as a reference for every calculation
+    // below to minimize storage access calls.
+    bytes32 _messageHash = keccak256(_proofs[0].message);
+    // TODO: Could use an array of sharedRoots so you can submit a message batch of messages with
+    // different origins.
+    bytes32 _inboundRoot = calculateMessageRoot(_messageHash, _proofs[0].path, _proofs[0].index);
+
+    // Check to see if the root for this batch has already been proven.
+    if (!provenRoots[_inboundRoot]) {
+      // TODO: If the updateAggregateRoot function hasnt been called for a while, we might be able
+      // to prove against the pending root which is more recent.
+      // TODO: Shouldn't we go ahead and move the pending aggregate root into current, in this case?
```

Comment:
yes there should be a helper that just does "getAggregateRoot()" that everything uses and it will upgrade current and return whatever is latest

Reply:
should that be considered in-scope here? It's a lil bit more complex than just adding a helper - see https://discord.com/channels/992549151134982234/1022516954986184794/1024079262489247917

Worth making a smol separate issue imo

Reply:
okay lets keep it out of scope of this PR -- can you make the issue and drop link here for continuity?

Reply:
https://github.com/connext/nxtp/issues/1963